### ISSUE TITLE
crash due profiling java application (s) on aarch64

### ISSUE THREAD

**Issue by rkgithubs** [DynamoRIO/dynamorio#3586]
I'm seeing the following crashes on aarch64 while running a couple of java applications, this happens to on JDK8 and JDK9 both. 
First one is SPECjvm08 benchmark, second one is a physics game test.
~/workloads/SPECjvm2008/SPECjvm2008_v101$ ~/tools/DynamoRIO/DynamoRIO-x86_64-Linux-7.90.17998-0/bin64/drrun -t drcachesim -offline -- ~/tools/jdk1.9-b160/bin/java -jar SPECjvm2008.jar scimark.sor.small -ikv -wt 30 -it 60
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007f9541ca5383, pid=29922, tid=29923
#
# JRE version:  (9.0+160) (build )
# Java VM: Java HotSpot(TM) 64-Bit Server VM (9-ea+160, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# v  ~BufferBlob::Signature Handler Temp Buffer
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/daenerys/workloads/SPECjvm2008/SPECjvm2008_v101/hs_err_pid29922.log
Loaded disassembler from /home/daenerys/tools/jdk1.9-b160/lib/server/hsdis-amd64.so
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#

~/workloads/SPECjvm2008/SPECjvm2008_v101$ ~/tools/DynamoRIO/DynamoRIO-x86_64-Linux-7.90.17998-0/bin64/drrun -t drcachesim -offline -- ~/tools/jdk1.9-b160/bin/java -jar ~/workloads/IcyRocks-Java-Kernel-V1.0.2.jar
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007f1c11ca5383, pid=29991, tid=29992
#
# JRE version:  (9.0+160) (build )
# Java VM: Java HotSpot(TM) 64-Bit Server VM (9-ea+160, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# v  ~BufferBlob::Signature Handler Temp Buffer
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/daenerys/workloads/SPECjvm2008/SPECjvm2008_v101/hs_err_pid29991.log
Loaded disassembler from /home/daenerys/tools/jdk1.9-b160/lib/server/hsdis-amd64.so
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
Segmentation fault

**Comment by rkgithubs** [comment]
I see #2506 from 2017 with similar issue on java and node apps.

**Comment by AssadHashmi** [comment]
This crash looks different from the one in #2506 in that this is a SIGILL.
Perhaps I'm not interpreting the crash information correctly but it looks like these are x86 binaries?
```
~/tools/DynamoRIO/DynamoRIO-x86_64-Linux-7.90.17998-0/bin64/drrun . . .
. . .
Java VM: Java HotSpot(TM) 64-Bit Server VM (9-ea+160, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
. . .
Loaded disassembler from /home/daenerys/tools/jdk1.9-b160/lib/server/hsdis-amd64.so
```

**Comment by rkgithubs** [comment]
Following is another crash seen in the java app:
This app runs successfully using this JDK without DR. Also, attached the loglevel 2 info. 
 /root/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/drrun -debug -loglevel 2 -- /root/jdk1.9-b160/bin/java -jar /root/IcyRocks-Java-Kernel-V1.0.2.jar 
<log dir=/root/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/../logs/java.139682.00000000>
<Starting application /root/jdk1.9-b160/bin/java (139682)>
<Initial options = -no_dynamic_options -loglevel 2 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/root/DynamoRIO-x86_64-Linux-7.90.18019-0/lib64/debug/libdynamorio.so' 0x00007f1e221bc580
>
<(1+x) Handling our fault in a TRY at 0x00007f1e22405d7c>
<spurious rep/repne prefix @0x00007f1e21f79f96 (f2 41 ff e3): >
<Invalid opcode encountered>
<Application /root/jdk1.9-b160/bin/java (139682).  Internal Error: DynamoRIO debug check failure: /home/travis/build/DynamoRIO/dynamorio/core/translate.c:931 false
(Error occurred @18897 frags)
version 7.90.18019, custom build
-no_dynamic_options -loglevel 2 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct 
0x00007f1e022a0b40 0x00007f1e2225700e
0x00007f1e022a0d90 0x00007f1e2236d68d
0x00007f1e022a0e90 0x00007f1e2236f21c
0x00007f1e022a1920 0x00007f1e2236f79a
0x00007f1e022a1960 0x00007f1e223483ae
0x00007f1e022a19c0 0x00007f1e2243c52f
0x00007f1e022a1ec0 0x00007f1e22441c1e
0x00007f1e022a21d0 0x00007f1e22444e38
0x00007f1e022a2370 0x00007f1e2240589d
0x00007f1e013251c0 0x49656e696c65746e>

[java.0.139682.txt](https://github.com/DynamoRIO/dynamorio/files/3179631/java.0.139682.txt)

**Comment by rkgithubs** [comment]
simple java -version failed. 
 /root/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/drrun -debug -loglevel 2 -- /root/jdk1.9-b160/bin/java -version    
<log dir=/root/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/../logs/java.139870.00000000>
<Starting application /root/jdk1.9-b160/bin/java (139870)>
<Initial options = -no_dynamic_options -loglevel 2 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/root/DynamoRIO-x86_64-Linux-7.90.18019-0/lib64/debug/libdynamorio.so' 0x00007f8f1f430580
>
<(1+x) Handling our fault in a TRY at 0x00007f8f1f679d7c>
<spurious rep/repne prefix @0x00007f8f1f1edf96 (f2 41 ff e3): >
<Invalid opcode encountered>
<Application /root/jdk1.9-b160/bin/java (139870).  Internal Error: DynamoRIO debug check failure: /home/travis/build/DynamoRIO/dynamorio/core/translate.c:931 false
(Error occurred @18379 frags)
version 7.90.18019, custom build
-no_dynamic_options -loglevel 2 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct 
0x00007f8eff4d7b40 0x00007f8f1f4cb00e
0x00007f8eff4d7d90 0x00007f8f1f5e168d
0x00007f8eff4d7e90 0x00007f8f1f5e321c
0x00007f8eff4d8920 0x00007f8f1f5e379a
0x00007f8eff4d8960 0x00007f8f1f5bc3ae
0x00007f8eff4d89c0 0x00007f8f1f6b052f
0x00007f8eff4d8ec0 0x00007f8f1f6b5c1e
0x00007f8eff4d91d0 0x00007f8f1f6b8e38
0x00007f8eff4d9370 0x00007f8f1f67989d
0x00007f8efe5991c0 0x49656e696c65746e>

**Comment by AssadHashmi** [comment]
From the `java.0.139682.txt` log file supplied:
```
. . .
SYSLOG_WARNING: spurious rep/repne prefix @0x00007f1e21f79f96 (f2 41 ff e3):
. . .
SYSLOG_WARNING: Invalid opcode encountered                                       
SYSLOG_ERROR: Application /root/jdk1.9-b160/bin/java (139682).  Internal Error: DynamoRIO debug check failure: /home/travis/build/DynamoRIO/dynamorio/core/translate.c:931 false
(Error occurred @18897 frags)                                                    
version 7.90.18019, custom build
. . .
```

Which is consistent with the `SIGILL`.
Can you confirm that the title of this issue is inaccurate in that you're not running AArch64?

Did you build the `/root/jdk1.9-b160/bin/java` binary from source or is it a standard pre-built download?

**Comment by derekbruening** [comment]
Please look in the thread log file (you've only supplied the global log file) to see what the invalid opcode is.  Is it AVX-512 (#1312) whose support is in-progress?

**Comment by rkgithubs** [comment]
#java -version shouldn't execute avx512 opcodes. This is jvm built from the source, also system is x86_64.

**Comment by hgreving2304** [comment]
Please check anyway.
>> This is jvm built from the source, also system is x86_64.
Could still be AVX-512. Have a look whether the instruction begins with
0x62.

*From: *rkgithubs <notifications@github.com>
*Date: *Thu, May 16, 2019 at 10:05 AM
*To: *DynamoRIO/dynamorio
*Cc: *Subscribed

#java -version shouldn't execute avx512 opcodes. This is jvm built from the

**Comment by rkgithubs** [comment]
Is there a recommended version of JDK on x64. Currently all java program crash with DR for JDK9+

**Comment by rkgithubs** [comment]
with jdk8 debug level 3; java app is stuck <proactive reset @ fragments>>. I have tried running SPEC jvm2008 and some multi-threaded tests- applications don't run/start as expected. Please suggest if any more info is needed. This is x86_64 and jdk8 built from source. 

~/rahul/java_tests# /root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/drrun -debug -loglevel 3 -- /root/rahul/jdk1.8.0_201/bin/java TestThread 
<log dir=/root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/../logs/java.186088.00000000>
<Starting application /root/rahul/jdk1.8.0_201/bin/java (186088)>
<Initial options = -no_dynamic_options -loglevel 3 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/lib64/debug/libdynamorio.so' 0x00007f106bec3580
>
<curiosity: rex.w on OPSZ_6_irex10_short4!>
<(1+x) Handling our fault in a TRY at 0x00007f106c10cd7c>
<spurious rep/repne prefix @0x00007f106bc80f96 (f2 41 ff e3): >
<writing to executable region.>
<get_memory_info mismatch! (can happen if os combines entries in /proc/pid/maps)
        os says: 0x00007f0fbc000000-0x00007f0fe8000000 prot=0x00000000
        cache says: 0x00007f0fbc000000-0x00007f0fc8000000 prot=0x00000000
>
<proactive reset @ 18507 fragments>
<proactive reset @ 41493 fragments>
<proactive reset @ 53141 fragments>
<proactive reset @ 60504 fragments>
<proactive reset @ 81809 fragments>

**Comment by rkgithubs** [comment]
top of stack trace 

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (sharedRuntime.cpp:549), pid=186643, tid=0x00007f78d61f3700
#  guarantee(cb != NULL && cb->is_nmethod()) failed: safepoint polling: pc must refer to an nmethod
#
# JRE version: Java(TM) SE Runtime Environment (8.0_201-b09) (build 1.8.0_201-b09)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.201-b09 mixed mode linux-amd64 compressed oops)
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#

---------------  T H R E A D  ---------------

Current thread (0x00007f79f8173000):  JavaThread "3" [_thread_in_Java, id=186719, stack(0x00007f78d60f3000,0x00007f78d61f4000)]

Stack: [0x00007f78d60f3000,0x00007f78d61f4000],  sp=0x00007f78d61ef020,  free space=1008k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0xad3525]  VMError::report_and_die()+0x2e5
V  [libjvm.so+0x4e0368]  report_vm_error(char const*, int, char const*, char const*)+0x58
V  [libjvm.so+0x9c071f]  SharedRuntime::get_poll_stub(unsigned char*)+0x6f
V  [libjvm.so+0x9190fc]  JVM_handle_linux_signal+0x8ac
V  [libjvm.so+0x90b8b8]  signalHandler(int, siginfo*, void*)+0x38
C  [libpthread.so.0+0x11390]
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+46
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30
j  MTDemo.fib(Ljava/math/BigInteger;)Ljava/math/BigInteger;+30

**Comment by derekbruening** [comment]
We are still waiting for info about the invalid opcode...note that Status-NeedInfo label.

**Comment by rkgithubs** [comment]
I'm not see invalid opcode with jdk8. but there is crash still. invalid opcode happens in later versions of jdk.

**Comment by rkgithubs** [comment]
The program crashes on jdk9 due to invalid opcode. log attached 

SYSLOG_WARNING: Invalid opcode encountered
Error decoding 0x00007f54d97fb4a5 == 0x62 0xf1 0x7d 0x8 0x6e 0xc1 
Error decoding 0x00007f54d97fb4a5 == 0x62 0xf1 0x7d 0x8 0x6e 0xc1 
Error decoding 0x00007f54d97fb4a5 == 0x62 0xf1 0x7d 0x8 0x6e 0xc1 
ERROR: recreate_app : looking for 0x0000000044a2db9f in frag @ 0x0000000044a2db90 (tag 0x00007f54d97fb4a5)
TAG  0x0000000000000000
 +0    L4 @0x00007f54fb7c1ea0  e9 a0 d7 14 de       jmp    $0x00007f54d97fb4a5
END 0x0000000000000000
[log.1.191280.txt](https://github.com/DynamoRIO/dynamorio/files/3209653/log.1.191280.txt)

**Comment by rkgithubs** [comment]
although i'm not sure what is happening with jdk8. there is no crash; what should i look for in the log so I'm attaching log for java process. 
[java.0.191566.txt](https://github.com/DynamoRIO/dynamorio/files/3209663/java.0.191566.txt)

Program is stuck here (based on debug loglevel 1)

 :~/rahul/java_tests# /root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/drrun -debug -loglevel 1  -- /root/rahul/jdk1.8.0_201/bin/java /root/rahul/java_tests/TestThread 
<log dir=/root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/../logs/java.191566.00000000>
<Starting application /root/rahul/jdk1.8.0_201/bin/java (191566)>
<Initial options = -no_dynamic_options -loglevel 1 -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/lib64/debug/libdynamorio.so' 0x00007ff92e756580
>
<(1+x) Handling our fault in a TRY at 0x00007ff92e99fd7c>
<spurious rep/repne prefix @0x00007ff92e513f96 (f2 41 ff e3): >
<writing to executable region.>
<get_memory_info mismatch! (can happen if os combines entries in /proc/pid/maps)
        os says: 0x00007ff8e0000000-0x00007ff8ec000000 prot=0x00000000
        cache says: 0x00007ff8e8000000-0x00007ff8ec000000 prot=0x00000000
>
<proactive reset @ 18456 fragments>
<proactive reset @ 41352 fragments>
<proactive reset @ 53081 fragments>
<proactive reset @ 53763 fragments>
<proactive reset @ 60430 fragments>
<proactive reset @ 67218 fragments>
<proactive reset @ 67419 fragments>

**Comment by derekbruening** [comment]
> The program crashes on jdk9 due to invalid opcode. log attached
> 
> SYSLOG_WARNING: Invalid opcode encountered
> Error decoding 0x00007f54d97fb4a5 == 0x62 0xf1 0x7d 0x8 0x6e 0xc1
> Error decoding 0x00007f54d97fb4a5 == 0x62 0xf1 0x7d 0x8 0x6e 0xc1
> Error decoding 0x00007f54d97fb4a5 == 0x62 0xf1 0x7d 0x8 0x6e 0xc1

That is AVX-512.

**Comment by derekbruening** [comment]
The wherewasi assert is #3175.  It's not serious -- maybe re-run with `-ignore_assert_list '*'` to let it continue to see if more significant things show up?

**Comment by derekbruening** [comment]
```
<proactive reset @ 18507 fragments>
<proactive reset @ 41493 fragments>
<proactive reset @ 53141 fragments>
<proactive reset @ 60504 fragments>
<proactive reset @ 81809 fragments>
```

These seem surprisingly early and frequent: that much memory shouldn't be used so quickly.  I would try with `-vm_size 1G -no_enable_reset -disable_traces`.  Maybe an experimental run which adds `-no_hw_cache_consistency -no_sandbox_writes`.

**Comment by rkgithubs** [comment]
application can be run via DR after adding -disable_traces. Wouldn't it disable the trace collection completely though. Our goal is to run drcachesim tool to collect the memory access instructions. 

/rahul/java_tests# /root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/drrun -s 60 -vm_size 1G -no_enable_reset -disable_traces  -- /root/rahul/jdk1.8.0_201/bin/java TestThread 
Creating 1
Starting 1
Creating 2
Starting 2
Running1
Running2
Thread: 2,125
Thread: 1,125

**Comment by derekbruening** [comment]
The DR core's traces have no relation to drcachesim's memory access traces.  See the `-disable_traces` docs http://dynamorio.org/docs/using.html#sec_options, see the DR tutorial slides, other DR documentation.

**Comment by rkgithubs** [comment]
once we tried to collect traces -t drachesim -offline trace collection failed with the following: 
hs_err_pid198432.log indicates the following exception thrown. 
Internal exceptions (2 events):
Event: 2.043 Thread 0x00007f1b2800a800 Exception <a 'java/lang/NoSuchMethodError': Method sun.misc.Unsafe.defineClass(Ljava/lang/String;[BII)Ljava/lang/Class; name or signature does not match> (0x00007f1899107cb0) thrown at [/HUDSON/workspace/8-2-build-linux-amd64/jdk8u201/12317/hotspot/src/sh
Event: 2.046 Thread 0x00007f1b2800a800 Exception <a 'java/lang/NoSuchMethodError': Method sun.misc.Unsafe.prefetchRead(Ljava/lang/Object;J)V name or signature does not match> (0x00007f1899107f98) thrown at [/HUDSON/workspace/8-2-build-linux-amd64/jdk8u201/12317/hotspot/src/share/vm/prims/jni.c


/root/rahul/DynamoRIO-x86_64-Linux-7.90.18019-0/bin64/drrun -vm_size
o_enable_reset -disable_traces -t drcachesim -offline -trace_after_instrs 5000M -exit_after_tracing 
-- /root/rahul/jdk1.8.0_201/bin/java TestThread
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f1b30370481, pid=198432, tid=0x00007f1246d30700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_201-b09) (build 1.8.0_201-b09)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.201-b09 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.so+0x2dd481]  Assembler::jmp_literal(unsigned char*, RelocationHolder const&)+0x61
#
# Core dump written. Default location: /root/rahul/java_tests/core or core.198432
#
# An error report file with more information is saved as:
# /root/rahul/java_tests/hs_err_pid198432.log

[error occurred during error reporting , id 0xb]

#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp

**Comment by derekbruening** [comment]
(Triple backtics delimit code blocks to make it easier to read -- I added some)

**Comment by derekbruening** [comment]
So the status is that jdk8 on even small Java apps seems to hit an app crash when run under plain DR with traces enabled, or when run under drcachesim -offline?  Likely both are related to cache consistency: probably the same or similar to #2989.

**Comment by rkgithubs** [comment]
crash is seen with drcachesim -offline on small apps even.

Rahul

**Comment by derekbruening** [comment]
> crash is seen with drcachesim -offline on small apps even

That is surprising given regression tests and usage elsewhere.  This sounds like a separate issue from this issue here about JDK8.  Please file a separate issue and provide the app where this happens.
